components/script/dom/blob.rs

```diff
@@ -118,35 +152,70 @@ impl Blob {
         // TODO: accept other blobParts types - ArrayBuffer or ArrayBufferView
         let bytes: Vec<u8> = match blobParts {
             None => Vec::new(),
-            Some(blobparts) => blob_parts_to_bytes(blobparts),
+            Some(blobparts) => match blob_parts_to_bytes(blobparts) {
+                Ok(bytes) => bytes,
+                Err(_) => return Err(Error::InvalidCharacter),
+            }
         };
 
-        let slice = DataSlice::new(Arc::new(bytes), None, None);
-        Ok(Blob::new(global, slice, &blobPropertyBag.get_typestring()))
+        let slice = DataSlice::from_bytes(bytes);
+        Ok(Blob::new(global, BlobImpl::new_from_slice(slice), &blobPropertyBag.get_typestring()))
     }
 
-    pub fn get_data(&self) -> &DataSlice {
-        &self.data
+    /// Get a slice to inner data, this might incur synchronous read and caching
+    /// XXX: it is still possible to bring down the whole program, but the API might be
+    ///      intuitive to use (since we read FS and cache in a transparent style)
+    pub fn get_slice(&self) -> DataSlice {
+        match self.blob_impl {
+            BlobImpl::File(ref id, ref slice) => {
+                match *slice.borrow() {
+                    Some(ref s) => s.clone(),
+                    None => {
+                        let global = self.global();
+                        let s = read_file(global.r(), id.clone()).expect("Read file in get_slice failed");
```

Comment:
This should return a Result, too, and callers should pretend that there are no bytes in the Err case.